### PR TITLE
Try a backup copy if the downloaded cutechess-cli is not working.

### PR DESCRIPTION
**Pull request by vdbergh**
Logic:

1) Check if cutechess-cli exists and is working.

2) If not: download it.

3) Check if downloaded cutechess-cli is working.

4) If not: try to restore a backup copy.

5) If successful check if the backup copy is working.

6) If not: bail out.

Also: do not treat MacOS specially in the updater.

Also: only setup/verify during worker startup.

Solves #1368 .

### CODE REVIEW

**Comment by vdbergh** [comment]
@ppigazzini I wonder if you could have a look at this PR? I think it is a nice refactoring of the setup of cutechess-cli. Perhaps it might also deal correctly with this issue (if it is solvable at all):

https://tests.stockfishchess.org/actions?action=&user=technologov&before=1657958686.643&count=100

File: worker/worker.py [hunk 1]
```diff
@@ -270,6 +277,130 @@ def get_credentials(config, options, args):
...
+    if not verify_required_cutechess(testing_dir, cutechess):
+        if len(EXE_SUFFIX) > 0:
+            zipball = "cutechess-cli-win.zip"
+        elif IS_MACOS:
+            zipball = "cutechess-cli-macos-64bit.zip"
+        else:
+            zipball = "cutechess-cli-linux-{}.zip".format(platform.architecture()[0])
+        try:
+            download_from_github(zipball, testing_dir)
+            zip_file = ZipFile(zipball)
```
Comment (ppigazzini) [inline worker/worker.py:355]: Context manager added in python 3.2, see https://docs.python.org/3.5/library/zipfile.html#zipfile.ZipFile

```python
            zip_file = ZipFile(zipball)
            zip_file.extractall()
            zip_file.close()
```

```python
            with ZipFile(zipball) as zip_file:
                zip_file.extractall()
```

Comment (vdbergh) [reply worker/worker.py]: Changed!

**Comment by vdbergh** [comment]
> If I'm not wrong the error messages are now only worker side.

Yes that is true. 

But IMHO checking that cutechess is working is part of the start up of the worker (like selecting a compiler).

**Comment by ppigazzini** [comment]
I meant that I don't view entries in Events log about cutechess-cli problems:
https://dfts-0.pigazzini.it/actions

**Comment by vdbergh** [comment]
The messages in the event log come from failed tasks. In this PR the check for a working cutechess-cli is done before a task is started. So there are no messages in the event log. 

My plan is to create a log api to post messages to the event log not associated with a task.

**Comment by ppigazzini** [comment]
A collateral problem found during the tests of this PR.
The worker update is checked/done in `fetch_and_handle_task()`, we have now some code than can prevent a fixing update.

If we add at example the support for ARM (eg the link to the `cutechess-cli-arm.zip`), an already installed worker will exit (because a not working cutechess) before checking the version.

**Comment by vdbergh** [comment]
I agree! So the update test should also be done a startup….

**Comment by vdbergh** [comment]
Latest push does an update test before checking cutechess.

**Review by ppigazzini — APPROVED** [review]

**Comment by ppigazzini** [comment]
Nice improvement, thank you @vdbergh :)
Triggered the workers update.

PS: please try to update `black`. Mine is latest v 22.6.0 and writes `2**30`, yours still writes `2 ** 30`, see
https://github.com/psf/black/pull/2726

**Pull request merged**